### ISSUE TITLE
Calendaring: Generate Suggested Term

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8620]
As a Court user, so that creating a new term is as easy as possible, I need DAWSON to use its data to provide me with suggested term calendar sessions in a way that allows collaboration to happen.

## Acceptance Criteria
* Petitions clerks, Case Services Supervisors and Clerk of the Court users can click a button to invoke the "Create Term" workflow
* Create Term workflow allows the user to input a trial term, name, and start/end date
* Output of the Create Term workflow is a downloaded spreadsheet (preference would be XLSX) that calculates the trial sessions that should be scheduled for the term, based on ruleset (see UX comments below for ruleset, see notes for example template)
  - Columns: weeks in the date range specified (labeled w/ date of the Monday of each week)
  - Rows: cities 
    - Final Row: No. of Sessions that week (totaling number of sessions in a given column)
  - Cells: Type of trial session (small, regular, hybrid) scheduled for that week
* Any already scheduled New/Calendared Special trial sessions are incorporated into the calculation/spreadsheet output 
* NO trial sessions are scheduled/created in this operation - this is strictly generation and download of a spreadsheet
* Spreadsheet filename = [TermName].whatever
 
## Notes
[Reminders.pdf](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNjFDQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--671778777672b91a6bef5a7d67d664b631e8348c/Reminders.pdf)

<img width="839" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNlltQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--8332edf52bf62b902b28911ee1d57d882481af2a/image.png" alt="image.png" />

## Story Definition of Ready (updated on 12/23/22)
The following criteria must be met in order for the user story to be picked up by the Flexion development team.
The user story must: 
- [ ] Is framed in business/user need, the value has been addressed.
- [ ] Includes acceptance criteria
- [ ] Has been refined
- [ ] Pre conditions have been satisfied. 

**Process:** 
Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (“standard testing”). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 5-19-22)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios have been refined to meet all acceptance criteria 
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [ ] Automated test scripts have been written, including visual tests for newly added PDFs.
 - [ ] Field level and page level validation errors (front-end and server-side) integrated and functioning.
 - [ ] Verify that language for docket record for internal users and external users is identical.
 - [ ] New screens have been added to pa11y scripts.
 - [ ] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html.
 - [ ] READMEs, other appropriate docs, and swagger/APIs fully updated.
 - [ ] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required).
 - [ ] Interactors should validate entities before calling persistence methods.
 - [ ] Code refactored for clarity and to remove any known technical debt.
 - [ ] If new docket entries have been added as seed data to `efcms-local.json`, 3 local s3 files corresponding to that docketEntryId have been added to `web-api/storage/fixtures/s3/noop-documents-local-us-east-1`
 - [ ] Acceptance criteria for the story has been met.
 - [ ] If there are special instructions in order to deploy into the next environment, add them as a comment in the story.
 - [ ] If the work completed for the story requires a reindex without a migration, or any other special deploy steps, apply these changes to the following flexion branches:
   - [ ] experimental1
   - [ ] experimental2
   - [ ] experimental3
   - [ ] experimental4
   - [ ] experimental5
   - [ ] experimental6
   - [ ] develop
 - [ ] Reviewed by UX on a deployed environment.
 - [ ] Reviewed by PO on a deployed environment. Can be deployed to the Court's `test` environment if prod-like data is required. Otherwise deployed to any `experimental` environment.
 - [ ] Deployed to the Court's `staging` environment.

**Comment by katiecissell** [comment]
UX Notes:

Please see [Figma file](https://www.figma.com/design/9vsM06sIFUsel7BR4xDI7X/Generate-Trial-Session-Calendar?node-id=411-9570&t=OTouaBo3fcAGdXyA-1) for full comments

A new link is added to the trial sessions page. It is visible regardless of what tab you are viewing. Only Petitions clerks, Case Services Supervisors and Clerk of the Court  will see the link.
<img width="1567" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeHV4QlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--0ba23ac4e473783079d84a88fde7e8c36b17a543/image.png" alt="image.png" />

When the link is clicked a modal appears.
<img width="878" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBL0wyQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--8cf626a9fde60bcd7ccb9aa07170037460439619/image.png" alt="image.png" />


When create term is clicked, Dawson will create a .xlsx or .csv file (depending on the engineers feedback when they pick this ticket up.) The file will create a calendar of the term and divide the eligible cases for each location into sessions. It will also respect additional rules outlined in the figma file.  A loading state will be used while Dawson creates this file. When its done, the file will be downloaded to the user's computer/browser.

No specials or hybrid S's will be generated by this calendar. (existing specials already scheduled will show)

Hardcoded Session creation rules:
<img width="1760" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBemNPQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--f9833810b28c7bddd35d3ac4ab0c43d694b23d44/image.png" alt="image.png" />

**Comment by ttlenard** [comment]
## Test Cases

#### 1) Petitions Clerk accesses the Trial Sessions Page; new link that says, "Create Term" displays to the left of the Trial Session Planning Report link at the top of the page.

- Log in as a Petitions Clerk
- Click on the Trial Sessions Header

Expected Results:
- [x] **NEW FUNCTIONALITY** - There is a new link for "Create Term" to the left of the Trial Session Planning report link.
- [x] **NEW FUNCTIONALITY** - Link is selected when a user tabs through the screen 



#### 2) Petitions Clerk clicks on the "Create Term" link; new modal displays.

- Continuing from the previous test
- Click on the Create Term link

Expected Results:
- [x] **NEW FUNCTIONALITY** - A new modal displays that matches the mockup
- [x] **NEW FUNCTIONALITY** - Clicking close in the upper right closes the modal and the user is navigated back to the Trial Sessions Screen
- [x] **NEW FUNCTIONALITY** - Clicking on the Cancel button closes the modal and the user is navigated back to the Trial Sessions Screen.
- [x] **NEW FUNCTIONALITY** - Clicking on the Create Term button without entering in any data displays validation errors to the user (see mocks).
- [ ] **NEW FUNCTIONALITY** - Clicking on the Trial Term dropdown displays the next three terms (terms should not be in the past)
- [x] **NEW FUNCTIONALITY** - Clicking on the calendar icon displays a calendar picker; user is unable to select a date in the past on the picker.
- [x] **NEW FUNCTIONALITY** - If a user manually inputs a date in the past, a validation error is displayed.
- [x] **NEW FUNCTIONALITY** - If a user manually types in a date in an incorrect format (not MM/DD/YYYY), a validation error is displayed



#### 3) Petitions Clerk adds in valid data into the modal; Clicks Create Term button; User sees a loading screen.

- Continuing from the previous test
- Input all valid data
- Click Create Term button

Expected Results:
- [x] **NEW FUNCTIONALITY** - Screen displays the loading screen to the user. (Dev discretion on if it is the async progress bar or the spinner on the screen)



#### 4) Loading screen completes; Document is downloaded to the user's computer.

- Continuing from the previous test
- Loading screen completes

Expected Results:
- [x] **NEW FUNCTIONALITY** - Windows explorer (or similar) displays for the user to select where on their device that they would like to save a xlsx document (dev discretion on if it is xlsx or csv).
- [x] **NEW FUNCTIONALITY** - File name for the document is [TermName].xlxs



#### 5) Petitions Clerk opens the file; reviews generated Spreadsheet of the Term that was created.

- Continuing from the previous test
- Open the document and review the data

Expected Results:
- [x] **NEW FUNCTIONALITY** - Top row displays "Week Of"
- [x] **NEW FUNCTIONALITY** - Next row is the header row that includes "City" in the first column, followed by a horizontal list of weeks (labeled with the date of the Monday of each week) depending on the term dates that were input in the modal
- [x] **NEW FUNCTIONALITY** - First column includes a list of Trial cities in alphabetical order
- [x] **NEW FUNCTIONALITY** - Last row is titled No. of Sessions, and includes the total number of sessions in each column
- [x] **NEW FUNCTIONALITY** - Generated sessions for regular display in Green
- [x] **NEW FUNCTIONALITY** - Generated sessions for small display in Blue
- [ ] **NEW FUNCTIONALITY** - Generated sessions for Hybrid display in Purple
- [x] **NEW FUNCTIONALITY** - Any already calendared Special sessions that fall within the selected Term are displayed on the correct week and are Orange.



#### 6) Petitions clerk pulls a Trial Session Planning Report; Using the data from that report, compare the results in the Generated Suggested Term spreadsheet with the data on the planning report.

- Continuing from the previous test
- Review the suggested sessions for each week/city based on the eligible Regular and small cases, and if sessions have been scheduled at the location in the previous terms.

Expected Results:
- [ ] **NEW FUNCTIONALITY** - For all weeks in the term, there is a maximum of 6 sessions per week (this includes any already scheduled special sessions)
- [ ] **NEW FUNCTIONALITY** - For any location, there is a maximum of 5 sessions per location (this includes any already scheduled special sessions)
- [ ] **NEW FUNCTIONALITY** - For a location that has less than 40 regular cases, a regular session is NOT created
- [ ] **NEW FUNCTIONALITY** - For a location that has 40 or more regular cases, a regular session may be created
- [ ] **NEW FUNCTIONALITY** - For a location that has more than 140 regular cases, 2 regular sessions may be created
- [ ] **NEW FUNCTIONALITY** - For a location that has less than 40 small cases, a small session is NOT created
- [ ] **NEW FUNCTIONALITY** - For a location that has 40 or more small cases, a small session may be created
- [ ] **NEW FUNCTIONALITY** - For a location that has more than 165 small cases, 2 small sessions may be created
- [ ] **NEW FUNCTIONALITY** - For a location that has 40 or more regular cases, a regular session may be created
- [ ] **NEW FUNCTIONALITY** - For a location that has at least 50 small + regular cases combined, a hybrid session may be created
- [ ] **NEW FUNCTIONALITY** - For a location that has not had any sessions calendared in the previous two terms, any session type (small, reg, hybrid) must be created. This session will be created regardless of it there are a minimum number of cases.


*DEV QUESTION - If I generated multiple suggested terms, will there be a different result each time? Is there randomization involved in generating the data?
If yes, we will want to repeat test case 6 and 7 to ensure that the rules still apply even when the placement of suggested terms is random.



#### 7) Petitions Clerk unblocks a number of cases for a few locations; re-runs the trial session planning report; Generates a new suggested term; results in the spreadsheet reflect the changes and sessions are suggested appropriately.

- Continuing from the previous test
- Pull the blocked cases report
- Unblock a number of cases for a requested trial locations
- Next, generate a trial session planning report to see the changes in the eligible case numbers
- Next, Generate a suggested term
- Review the suggested sessions for each week/city based on the eligible regular and small cases, and if sessions have been scheduled at the location in the previous terms.

Expected Results
- [ ] **NEW FUNCTIONALITY** - All of the test cases from test case 6 pass
- [ ] **NEW FUNCTIONALITY** - The city that you unblocked multiple cases in may have a different result in the suggested term spreadsheet.

#### 8) Regression test to ensure that no sessions are actually created.

- As a petitions clerk, navigate to the Trial Sessions tab
- Review the existing sessions (new/open)

Expected Results:
- [ ] No new sessions were created when generated the suggested term spreadsheet.  



#### 9) Repeat all of the tests with the following users:
- [ ] Clerk of the Court
- [ ] Case Services Supervisor

**Comment by TomElliottFlexion** [comment]
New Rules: - Prioritize larger of small vs regular set of eligible cases
- Apply same rule to hybrid (leftover) sessions
- (Don't worry about it yet: Low volume cities should not be prioritized)

**Comment by akuny** [comment]
Based on discussion with UX the afternoon of 9/23/24, moving forward we will not include the "trial term" select element in the form for the first iteration of the suggested term generation workflow. The term name field will be the filename of the resulting Excel document, and the start and end date fields will be used by the module that generates the schedule to determine when the term starts and ends.

**Comment by TomElliottFlexion** [comment]
Compiled notes from several discussions: 

- `SMALL_CITIES` do not, in fact, get any `regular` sessions
- Only `special` `calendared` trial sessions scheduled in advance are taken into consideration when running the generator
- When determining whether a location has been visited in the last two terms, `special` trial sessions do not count
- ~~There is no minimum number of cases~~ The minimum number of cases when scheduling a trial session in a city that has not been visited in two terms is one

**Comment by TomElliottFlexion** [comment]
- Motion/Hearing sessions should be ignored completely for the purposes of the generator

**Comment by TomElliottFlexion** [comment]
- Washington, DC has two courtrooms available and will be treated as two separate locations for the purposes of scheduling
  - At most one of these can be taken by a non-special session

- (These are now categorized as Washington (North) and Washington (South). North will only ever get Special sessions, and South will only get special sessions if two are scheduled for the same week in Washington, DC)

**Comment by TomElliottFlexion** [comment]
- If a location has ZERO cases, it should not get a trial session regardless of whether or not it has been visited in the last two terms

**Comment by akuny** [comment]
Latest feedback:

1. The error message stating that there are more than two special trial sessions scheduled for Washington DC in the same week should be changed to a warning message, and the spreadsheet should generate anyhow.
2. In the spreadsheet, the cells containing special sessions on the triple-booked week should be red with an asterisk to call attention to the issue
3. Add a second worksheet to the spreadsheet that lists the procedure type/location mismatches that currently appear in the success message. (If it's not feasible to add a second worksheet, then list these mismatches below the calendar on the first worksheet.)

**Comment by katiecissell** [comment]
- [x]  Typing in dates in the past shows wrong validation message for start date. Should say something like dates can’t be in the past, I think we have a message like this somewhere already?   
- [x]  Term end date is not catching its in the past as long as its later than the start date

![Image](https://github.com/user-attachments/assets/1c101f0f-922f-43e8-96e3-895c7c9cdd28)

- [x]  Very aggressive tooltip that says create term on pretty much everything. Also lowercase the T for term.

![Image](https://github.com/user-attachments/assets/0686fb5d-4fa3-458c-a074-987150b9a420)

- [x]  Really long term names confuse the validation instead of it telling me there should be a limit (I’m assuming there is one)
![Image](https://github.com/user-attachments/assets/92fc2aea-4fb4-452b-a62b-4ad22dcfe3c3)


- [x] Smaller screen sizes push the clear button in a weird spot. Try and  put Clear button in top right and let header wrap with a bit of padding. 
![Image](https://github.com/user-attachments/assets/8502db01-ddfd-4915-ae74-9c4fcc9f29ac)


- [x] update gray background for labels to #dcdee0
- [x] Update red background for triple booked specials to #b50909 and change font color to white
- [x] Is there any way to make the cell expand to the width of the words by default?
![Image](https://github.com/user-attachments/assets/d9bf171d-ca6f-4d60-9119-d147ca231d87)


- [x] Left align bottom totals and case totals and remaining cases numbers
- [x] Incorrectly sized cases - remove red background on tab
- [x] Add a “Warnings” tab and display a list of the warnings
- [x] What happens if we’re unable to generate the spreadsheet?

**Comment by akuny** [comment]
Below is a screenshot of the error message that will display if we are unable to generate the spreadsheet:

![Image](https://github.com/user-attachments/assets/33e45d13-e9cf-4af7-94aa-355288898ade)

**Comment by swongCO** [comment]
Tested colors for color blindness (protanopia and deuteranopia) and found issues with the current colors for Special and Hybrid. Screenshots below show Small / Special and Regular / Hybrid are not differentiated. 
![Image](https://github.com/user-attachments/assets/704c192b-2972-4618-95d2-4d35ddfcb77d)

Here are the new suggested colors:
#ffbe2e = Special
#fee685 = Hybrid

**Comment by katiecissell** [comment]
Updated warning messages:
Show errors once for Maximum sessions per week exceeded and Maximum sessions per location exceeded

Trigger: Maximum sessions per week exceeded
Updated Wording: More special sessions than maximum allowed per week: mm/dd
Highlight: nothing for now, future perhaps header

Trigger: One session per location per week exceeded
Updated: More than one special trial per week scheduled: Location, mm/dd.
Highlight: The offending Session

Trigger: Maximum sessions per location exceeded
Updated: More special sessions than maximum allowed per location scheduled: Location
Highlight: nothing for now, future perhaps header

Washington specific:

Trigger:Two SPECIAL sessions per week in Washington DC exceeded
Updated: More than two special trial sessions per week: Washington, DC, 2/10.
Highlight: The offending Session

Trigger: Maximum sessions per location for Washington DC (South courtroom) exceeded
Updated: More special sessions than maximum allowed per location scheduled: Washington, DC
Highlight: nothing for now, future perhaps header


If no warnings display in warning tab: There were no warnings for this calendar.

**Comment by akuny** [comment]
> If no warnings display in warning tab: There were no warnings for this calendar.

When there are no warnings from generating the term, the "Warnings" tab will not appear at all in the resulting spreadsheet.

**Comment by katiecissell** [comment]
- [x] Get rid of period in warning alert header:
 ![Image](https://github.com/user-attachments/assets/26e8b4de-2caa-454d-9d01-8d7082b8c2b4)
- [x] Update unsuccessful alert text:
**Unable to create term**
There are no trial sessions to schedule within the dates provided.
![Image](https://github.com/user-attachments/assets/94c504bd-dbc1-4f25-baa4-5cf620f861db)

**Comment by cholly75** [comment]
We are moving this to production to get better feedback with live data.
